src/CurveCallee.sol

```diff
@@ -0,0 +1,182 @@
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "ds-math-sub-underflow");
+    }
+    function divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = add(x, sub(y, 1)) / y;
+    }
+
+    constructor(
+        address curveAddr_,
+        address uniV3Addr_,
+        address daiJoinAddr_
+    ) public {
+        curve = CurveLike(curveAddr_);
+        uniV3 = UniV3Like(uniV3Addr_);
+        daiJoin = DaiJoinLike(daiJoinAddr_);
+        dai = daiJoin.dai();
+
+        dai.approve(daiJoinAddr_, uint256(-1));
+    }
+
+    receive() external payable {}
+
+    function _fromWad(address gemJoin, uint256 wad) internal view returns (uint256 amt) {
+        amt = wad / 10 ** (sub(18, GemJoinLike(gemJoin).dec()));
+    }
+
+    function clipperCall(
+        address sender,            // Clipper caller, pays back the loan
+        uint256 owe,               // Dai amount to pay back        [rad]
+        uint256 slice,             // Gem amount received           [wad]
+        bytes calldata data        // Extra data, see below
+    ) external {
+        (
+            address to,            // address to send remaining DAI to
+            address gemJoin,       // gemJoin adapter address
+            uint256 minProfit,     // minimum profit in DAI to make [wad]
+             uint24 poolFee,       // uniswap V3 WETH-DAI pool fee
+            address charterManager // pass address(0) if no manager
+        ) = abi.decode(data, (address, address, uint256, uint24, address));
+
+        address gem = GemJoinLike(gemJoin).gem();
+        require(keccak256(bytes(TokenLike(gem).symbol())) == keccak256("wstETH"), "CurveCallee: only-wsteth");
+
+        // Convert slice to token precision
+        slice = _fromWad(gemJoin, slice);
+
+        // Exit gem to token
+        if(charterManager != address(0)) {
+            CharterManagerLike(charterManager).exit(gemJoin, address(this), slice);
+        } else {
+            GemJoinLike(gemJoin).exit(address(this), slice);
+        }
+
+        slice = WstEthLike(gem).unwrap(slice);
+        gem = WstEthLike(gem).stETH();
+
+        TokenLike(gem).approve(address(curve), slice);
+        slice = curve.exchange({
+                 i: 1,     // send token id 1 (stETH)
+                 j: 0,     // receive token id 0 (ETH)
+                dx: slice, // send `slice` amount of stETH
+            min_dy: 0      // accept any amount of ETH (`minProfit` is checked below)
+        });
+
+        gem = uniV3.WETH9();
+        WethLike(gem).deposit{
+            value: slice
+        }();
+
+        // Approve uniV3 to take gem
+        WethLike(gem).approve(address(uniV3), slice);
+
+        // Calculate amount of DAI to Join (as erc20 WAD value)
+        uint256 daiToJoin = divup(owe, RAY);
+
+        // Do operation and get dai amount bought (checking the profit is achieved)
+        bytes memory path = abi.encodePacked(gem, poolFee, address(dai));
```

Comment:
Why not get the uniswap path as input instead of encoding it on chain? Similar to the uniswap callees.

Reply:
Because in this case it will always be the same, so no need to get it as a parameter.

Reply:
You are right, `poolFee` may change as there are at least two ETH/DAI pools on UniV3. That is why it is a parameter. I still don’t get why you want to avoid doing the encoding on-chain.

Reply:
To save the gas of course.
We already get a Univ3 path in UniV3Calle.sol, so no reason for it to be different.

Reply:
This is a memory opération. The amount of gas saved is probably négligible. It does make sense for it to be différent, since this is a contract spécifique for wstETH. Asking the user to provide a path, but silently failing if they don’t provide a spécifique path seems counterintuitive to me.